src/librustc/dep_graph/dep_tracking_map.rs

```diff
@@ -134,7 +175,7 @@ impl<M: DepTrackingMapConfig> MemoizationMap for RefCell<DepTrackingMap<M>> {
     /// accesses the body of the item `item`, so we register a read
     /// from `Hir(item_def_id)`.
     fn memoize<OP>(&self, key: M::Key, op: OP) -> M::Value
-        where OP: FnOnce() -> M::Value
+        where OP: FnOnce() -> Option<M::Value>
```

Comment:
Ah, I thought @eddyb refactored that use. I wonder if we want to do so in this PR instead. Right now what happens is that we request `AssociatedItem(X)`, which is some summary information about some item X that is part of an impl or trait:

```rust
impl Foo {
    type A;
    type X;
    fn b() { ... }
}
```

To do this, we go up to the enclosing impl (the one shown there) and then iterate over its list of items. Right now, since we are iterating anyway, we would compute the associated item info for `A` and `b()` while we're at it (to avoid an `O(n^2)` blow-up, basically). We have been talking though about trying to refactor into finer-grained tasks that don't have this idea of generating "N results" from one execution. 

We could thus change this code to not generate N outputs but just find the one item we actually want. It would be O(N^2) but N here is typically very small.

And we do plan to rework things over time that would mitigate -- for example, moving to a newer system where we wouldn't have to be so careful about our dependencies. The whole motivation here was to only read from the impl when computing this data, and *not* the items themselves. This way, if the item X changes, we don't have to consider the value as dirty (the parts of X that would affect this value are copied into the impl and hashed there too). But we want to move to a system where we *would* recompute this value (which is cheap) and then see that, indeed, it has not changed.